src/unix/apple/audio.c

```diff
@@ -31,46 +30,188 @@ static void audio_queue_callback(void *opaque, AudioQueueRef q, AudioQueueBuffer
-	const char *deviceID, bool fallback)
+static OSStatus audio_object_get_device_uid(AudioObjectID device, AudioObjectPropertyScope prop_scope,
+	AudioObjectPropertyElement prop_element, char **out_uid, CFStringRef *out_uid_cf)
 {
-	// TODO Should this use the current run loop rather than internal threading?
+	char *uid = NULL;
+	CFStringRef uid_cf = NULL;
+
+	UInt32 data_size = sizeof(CFStringRef);
+	AudioObjectPropertyAddress propAddr = {
+		.mSelector = kAudioDevicePropertyDeviceUID,
+		.mScope = prop_scope,
+		.mElement = prop_element,
+	};
+
+	OSStatus e = AudioObjectGetPropertyData(device, &propAddr, 0, NULL, &data_size, &uid_cf);
+	if (AUDIO_HW_ERROR(e))
+		goto except;
 
-	MTY_Audio *ctx = MTY_Alloc(1, sizeof(MTY_Audio));
-	ctx->sample_rate = sampleRate;
-	ctx->channels = channels;
+	UInt32 prop_size = CFStringGetMaximumSizeForEncoding(CFStringGetLength(uid_cf), kCFStringEncodingUTF8);
+	uid = calloc(1, prop_size + 1);
+	if (!CFStringGetCString(uid_cf, uid, prop_size + 1, kCFStringEncodingUTF8)) {
+		e = kAudioHardwareBadDeviceError;
+		goto except;
+	}
+
+	// OK
+	*out_uid = uid;
+	*out_uid_cf = uid_cf;
+
+	except:
+
+	if (AUDIO_HW_ERROR(e)) {
+		if (uid_cf)
+			CFRelease(uid_cf);
+		free(uid);
+	}
+
+	return e;
+}
+
+static OSStatus audio_device_create(MTY_Audio *ctx, const char *deviceID)
+{
+	AudioObjectID *selected_device = NULL;
+	CFStringRef selected_device_uid = NULL;
+
+	// Enumerate all output devices and identify the given device
+	AudioObjectID *device_ids = NULL;
+
+	bool default_dev = !deviceID || !deviceID[0];
+
+	AudioObjectPropertyAddress propAddr = {
+		.mSelector = default_dev ? kAudioHardwarePropertyDefaultOutputDevice : kAudioHardwarePropertyDevices,
+		.mScope = kAudioObjectPropertyScopeOutput,
+		.mElement = kAudioObjectPropertyElementWildcard,
+	};
+
+	UInt32 data_size = 0;
+	OSStatus e = AudioObjectGetPropertyDataSize(kAudioObjectSystemObject, &propAddr, 0, NULL, &data_size);
+	if (AUDIO_HW_ERROR(e))
+		goto except;
+
+	device_ids = calloc(1, data_size);
+	e = AudioObjectGetPropertyData(kAudioObjectSystemObject, &propAddr, 0, NULL, &data_size, device_ids);
+	if (AUDIO_HW_ERROR(e))
+		goto except;
 
-	uint32_t frames_per_ms = lrint((float) sampleRate / 1000.0f);
-	ctx->min_buffer = minBuffer * frames_per_ms;
-	ctx->max_buffer = maxBuffer * frames_per_ms;
+	uint32_t n = (uint32_t) (data_size / sizeof(AudioObjectID));
+
+	if (default_dev) {
+		if (n != 1) {
+			e = kAudioHardwareBadDeviceError;
+			goto except;
+		}
+
+		selected_device = device_ids;
+
+	} else {
+		// Goal: find the AudioObjectID of the device whose unique string ID matches the given `deviceID` parameter
+		for (uint32_t i = 0; !selected_device && i < n; i++) {
+			char *uid = NULL;
+			CFStringRef uid_cf = NULL;
+
+			if (AUDIO_HW_ERROR(audio_object_get_device_uid(device_ids[i],
+				propAddr.mScope, propAddr.mElement, &uid, &uid_cf)))
+				continue;
+
+			if (!strcmp(deviceID, uid)) {
+				selected_device = &device_ids[i];
```

Comment:
```suggestion
				selected_device = device_ids + i;
```

Okay to leave as is.